### ISSUE TITLE
ZRPL (Seller App)- Compliance check

### ISSUE THREAD

**Issue by sahil-ondc** [ONDC-Official/v1.2.0-logs#211]
**search_full_catalog_refresh**

- /context must have required property 'ttl'
- /context/action must be equal to constant (search)

**on_search_full_catalog_refresh**

- /message/catalog/bpp/providers/0 must have required property 'fulfillments'
- /message/catalog/bpp/providers/items/quantity/available/count must be equal to one of the allowed values (99,0)
- /message/catalog/bpp/providers/0/items/16/price/value must be string
- /message/catalog/bpp/providers/0/items/16/price/maximum_value must be string

**on_search_inc_refresh**

- /context/action must be equal to constant (on_search)
- provider id: P1 in bpp/providers didn't matched for providers id of /on_search_full_catalog_refresh

**on_select**

- /context/action must be equal to constant (on_select)
- /message/order/quote must have required property 'ttl'

**init**

- address.building should be more than 3 chars
- address.name should be more than 3 chars
- Timestamp for /on_select api cannot be greater than or equal to /init api

**on_init**

- /context/action must be equal to constant (on_init)"
- /message/order/quote/breakup/2 must have required property '@ondc/org/item_id'
- Buyer app finder fee can't change in /on_init

**confirm**

- /message/order/quote/breakup/2 must have required property '@ondc/org/item_id

**on_confirm**

- /message/order must have required property 'tags'
- /message/order/fulfillments/0/end must have required property 'person'

**on_status**

- /context must have required property 'ttl'
- /message/order must have required property 'payment'
- /message/order/fulfillments/0 must have required property '@ondc/org/TAT'
- /message/order/quote must have required property 'ttl'
- order.updated_at timestamp should be less than or eqaul to context timestamp for /on_status API

@SudarshanSirsi

**Comment by sahil-ondc** [comment]
**on_search_inc_refresh**

- /context/action must be equal to constant (on_search)

**on_init**

- Buyer app finder fee can't change in /on_init

**on_confirm**

- /message/order must have required property 'tags'
- quote issues
  - "Key 'breakup.0.@ondc/org/item_id' mismatch: D560367WIN9S !== 154
  - Key 'breakup.1.@ondc/org/item_id' mismatch: C587505WIN8 !== 150
  
  @SudarshanSirsi

**Comment by sahil-ondc** [comment]
## Flow 1

**on_search_full_catalog_refresh**

- /message/catalog/bpp/providers/0/fulfillments/0/contact/phone must NOT have more than 11 characters
- category in serviceability construct should be one of the category ids bpp/providers[0]/items/category_id

## Flow 2

**on_confirm**

- /messgae/order/tags should have same and valid gst_number as passed in /confirm

## Flow 3

**on_select**

- only out of stock item/id should be provided in error/message

@SudarshanSirsi

**Comment by sahil-ondc** [comment]
## Flow 1

**on_search_full_catalog_refresh**

- remove formatting from /message/catalog/bpp/providers/items/descriptor
- /message/catalog/bpp/providers/items/time/timestamp cannot be greater than context/timestamp

**on_search_inc_catalog_refresh**

- /message/catalog/bpp/providers/items/time/timestamp should only change in case of change in /message/catalog/bpp/providers/items/time/label

## Flow 2

**on_select**

- available and maximum count should not change; the logic remains the same as in /on_search catalog

**on_status**

- /message/order/fulfillments/start/time/timestamp cannot match its range/end

## Flow 3

**on_select**

- available and maximum count should not change; the logic remains the same as in /on_search catalog
- item with id 'ICC-C583504WIN8' was in stock in on_search_full_catalog and no other inc_catalog is submitted for out of stock

## Flow 4

**on_cancel**

- Transaction logs should be as per the api contract

@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
Hi @sahil-ondc 
Flow 1:
on_search_full_catalog_refresh:
1. remove formatting from /message/catalog/bpp/providers/items/descriptor -> Can you please tell more about this, what do you 
    mean by remove formatting? 
2. /message/catalog/bpp/providers/items/time/timestamp cannot be greater than context/timestamp --> all the timestamps in  
    /message/catalog/bpp/providers/items/time/timestamp are lesser then the context.timestamp only, can you please tell me 
    which one you are poiniting to?

Flow 2: on_select
4.  available and maximum count should not change; the logic remains the same as in /on_search catalog --> So how do 
     buyer_app get to know the actual stock of items present in seller app inventory
5. /message/order/fulfillments/start/time/timestamp cannot match its range/end --> can you please explain more about this 
    one?as I understood the by API contract is that /message/order/fulfillments/start/time/timestamp refers to the expected 
    shipping timestamps and  /message/order/fulfillments/end/time/timestamp refers to the expected delivery timestamp, can 
   you please elaborate on what am I missing here and what needs to be done?

Flow 4: on_cancel :
6.  It is in compliance with the api contract as I checked, can you tell what exactly you are looking or what I have missed here?
  
   @sahil-ondc

**Comment by sahil-ondc** [comment]
**on_search_full_catalog_refresh**

1. Just remove the extra spaces in between the text and /n as well
2. Path is given for the timestamp which is greater then context/timestamp


**on_select**

3. for item if you have given available and maximum 99, 99 then it should be same in other calls as well, and vice versa

**on_status**

4. provided timestamp should be in the range as provided. either in start or end

 **on_cancel**

5. message/order/tags/quote_trail is missing

@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
Hi @sahil-ondc 
on_search_full_catalog_refresh:
2. currently the context.timestamp=2023-12-25T06:16:39.239Z and all the other message/catalog/bpp/providers/items/time/timestamp are lesser then the above context.timestamp can you please verify it once again?

**Comment by sahil-ondc** [comment]
## Flow 2

### on_init

- provider_location should not be provided in /message/order (Deprecated)
  `https://docs.google.com/document/d/1brvcltG_DagZ3kGr1ZZQk4hG4tze3zvcxmGV4NMTzr8/edit#heading=h.ks3uqu8tq83s`
- quote issues
  - Key length mismatch for breakup.2
  
  @SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
Hi @sahil-ondc for the next log resubmission do we need to submit for the flow 1 as well? since there is no changes in flow 1 can we skip that for next submission and only submit for other flows?

**Comment by bluecypher** [comment]
Yes, pls fix all the issues as mentioned in the feedback above.

@SudarshanSirsi

**Comment by bluecypher** [comment]
## Flow 1

**/on_search**

- "net_quantity_or_measure_of_commodity_in_pkg" and "imported_product_country_of_origin" properties are not required in /@ondc/org/statutory_reqs_packaged_commodities
- "storage_type" property is missing in /tags/attribute
- "storage" should ideally be in stringified integer format; buyer apps will concatenate storage, storage_unit and storage_type to render the details
- relevant values should be provided in brand/model for item id: ZA540019IN
- are variants not supported?

## Flow 2

**/on_confirm**

- estimated time ranges should be future dated; /(fulfillments/start/time/range/start can't be < context/timestamp)
- @ondc/org/TAT is missing in /fulfillments
- Is tracking supported? tracking is set as true;

**/on_status**

- "@ondc/org/provider_name" should not change
- message_id should be unique for every unsolicited call

## Flow 3

**/on_select (OOS)**

- error.message should be in stringified object notation as mentioned in the contract

## Flow 4

**/cancel**

- "012" is not a valid cancellation reason code; SNP must respond with a NACK

## Flow 5

- why are 2 on_status called for same status (Order-picked-up)
- Invalid RTO /on_cancel flow

## Flow 6

- flow must be provided for verification


@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
Hi @bluecypher 
Flow 1: on_search
5. variants are not supported at this time,

Flow 5:
1. we support only Accepted, In-progress, Completed. So once the order is picked up we try to deliver, but buyer was not available, hence attempted another time, so picked up 2 times. So If this implementation is not good can you tell us how we should proceed for this flow. can we do Order-picked-up (one time) and then seller cancels the product, is this enough?
2. We have off network logistic service, and we are an ISN, Is this mandatory to submit RTO on_cancel flow?
3. In our whole business model, we never support Return and refund and buyer cancel also. Suppose if we develop also, it will be of no use for us except for log verification. So is this really necessary for us?

**Comment by Jagannath-wits** [comment]
# Flow 1
### Full Catalog Refresh Errors

- **Schema Error**: Ensure the catalog includes the 'bpp/fulfillments' property.
- **Incorrect Tags for Electronics Items**:
 - Items 0 through 15 mandatory tag fields are incorrect and should be exactly:
    - Storage unit
    - Storage Type
    - Screen Size
    - OS Type
    - OS Version

### Incremental Refresh Errors

- **City Context**: Set 'context/city' to "*" in '/on_search_inc_refresh'.

Ensure all mandatory fields are filled out correctly. Refer to the [documentation](https://docs.google.com/spreadsheets/d/1APAvavF_BNbTA89benAlGtv0GuFvpn2b6XXi4lSdTTw/edit#gid=244000021) for guidance.


# Flow 2
### on_select

- **schemaErr0**: /message/order/provider must have required property 'locations'
- **schemaErr1**: /message/order/fulfillments/0/@ondc~1org~1TAT must match format "duration"
- **prvdrLoc**: provider.locations[0].id is missing in /on_select
- **ttstat**: /fulfillments/@ondc/org/TAT (O2D) in /on_select can't be less than @ondc/org/time_ship (O2S) in /on_search

### on_init

- **quoteErr0**: Mismatch found for item with item_id b452f4b8-6d11-43f0-ab70-61824dfbfcc8 while comparing quote object of on_select and on_init i.e please keep decimal in both calls

### confirm

- **quoteErr0**: Mismatch found for item with item_id b452f4b8-6d11-43f0-ab70-61824dfbfcc8 while comparing quote object of on_select and confirm i.e please keep decimal in both calls

### on_confirm

- **tmpstmp**: context/timestamp difference between /on_confirm and /confirm should be less than 5 sec
- **message.order.fulfillments[b452f4b8-6d11-43f0-ab70-61824dfbfcc8]**: 'TAT' must be provided in message/order/fulfillments
- **updtdtmstmp**: order.updated_at timestamp should be updated as per the context.timestamp (since default fulfillment state is added)
- **invldUpdtdTmstp**: updated_at timestamp should be equal to context timestamp for /on_confirm
- **quoteErr0**: Mismatch found for item with item_id b452f4b8-6d11-43f0-ab70-61824dfbfcc8 while comparing quote object of on_confirm and on_select i.e please keep decimal in both calls
- **message/order/tags/bpp_terms/np_type**: np_type is missing in message/order/tags/bpp_terms for ON_CONFIRM
- **message/order/tags/bpp_terms**: List of bpp_terms mismatched in message/order/tags/bpp_terms for on_init and on_confirm


Flow 4
### on_cancel

- **schemaErr0**: /context must have required property 'ttl'
- **message/order/transaction_id**: Transaction_id missing in message/order/payment
- **missingPreCancel**: Pre Cancel is mandatory for on_cancel

Flow 5

### on_cancel

- **schemaErr0**: /context must have required property 'city'
- **schemaErr1**: /message/order/fulfillments/1 must have required property 'id'
- **schemaErr2**: /message/order/fulfillments/1 must have required property 'state'
- **schemaErr3**: /message/order/fulfillments/1 must have required property 'type'
- **schemaErr4**: /message/order/fulfillments/1 must have required property 'tags'
- **schemaErr5**: /message/order/fulfillments/2/start/location must have required property 'id'
- **schemaErr6**: /message/order/fulfillments/2/start/location must have required property 'descriptor'
- **schemaErr7**: /message/order/fulfillments/2/end must have required property 'time'
- **schemaErr8**: /message/order/fulfillments/2/end must have required property 'person'
- **schemaErr9**: /message/order/fulfillments/2/end must have required property 'contact'
- **schemaErr10**: /message/order/fulfillments/2/end/location/address must have required property 'name'
- **schemaErr11**: /message/order/fulfillments/2/end/location/address must have required property 'building'
- **schemaErr12**: /message/order/fulfillments/2/end/location/address must have required property 'country'
- **RTO.start.location/DeliveryFulfillment.end.location**: RTO fulfillment start and Delivery fulfillment end location mismatch in on_cancel
- **RTO.end.location/DeliveryFulfillment.start.location**: RTO fulfillment end and Delivery fulfillment start location mismatch in on_cancel
- **missingPreCancel**: Pre Cancel is mandatory for on_cancel

Flow 6-a

### on_update

- **schemaErr0**: /message/order/quote must have required property 'ttl'
- **schemaErr1**: /message/order/payment/params must have required property 'transaction_id'
- **message/order/transaction_id**: Transaction_id missing in message/order/payment

# Flow 6-b
Please provide reverse qc as well as yours is RET14 Domain

# Flow 6-c

### on_update_interim

- **schemaErr0**: /message/order must have required property 'payment'
- **schemaErr1**: /message/order/quote must have required property 'ttl'
- **invldQuoteTrailPrices**: quote_trail price and item quote price sum for on_update should be equal to the price as in on_confirm

### on_update_liquidated
 
- same as above

@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
Hi @Jagannath-wits  expecting clarification for the following ASAP

Flow 1, Full catalog refresh

1. In the api contract for electronics there is no mention of bpp/fulfillments in catalog, and also it is said to be deprecated previously, and marked as to be deprecated for other categories as well
3. Items 0 through 15 mandatory tag fields are incorrect and should be exactly:
Storage unit
Storage Type
Screen Size
OS Type
OS Version

here what I see is that only storage type is missing , but I dont understood what should be done with other fields you mentioned. We have added those fields according to the api documents for 1.2 only

Flow 6: on_update
What about the reverse qc? because are registered as an ISN seller app with off network logistic we do not support returns and cancel in any time in our business in feature. In case of issue also, we provide replacement for the same product but not return. Since replacement flow is not there we took the liquidation flow, that too only for log validation purpose. So I dont understand where reverse qc part will come for liquidation and what is the necessity of the return considering above

**Comment by Jagannath-wits** [comment]
> Hi @Jagannath-wits expecting clarification for the following ASAP
> 
> Flow 1, Full catalog refresh
> 
> 1. In the api contract for electronics there is no mention of bpp/fulfillments in catalog, and also it is said to be deprecated previously, and marked as to be deprecated for other categories as well
> 2. Items 0 through 15 mandatory tag fields are incorrect and should be exactly:
>    Storage unit
>    Storage Type
>    Screen Size
>    OS Type
>    OS Version
> 
> here what I see is that only storage type is missing , but I dont understood what should be done with other fields you mentioned. We have added those fields according to the api documents for 1.2 only
> 
> Flow 6: on_update What about the reverse qc? because are registered as an ISN seller app with off network logistic we do not support returns and cancel in any time in our business in feature. In case of issue also, we provide replacement for the same product but not return. Since replacement flow is not there we took the liquidation flow, that too only for log validation purpose. So I dont understand where reverse qc part will come for liquidation and what is the necessity of the return considering above

Hi @SudarshanSirsi , 

-  bpp/fulfillments is mandatory for all domains.
- The other fields are correct please add the missing fields

**Comment by sahil-ondc** [comment]
ZRPL

# Flow 1

## on_search_full_catalog_refresh

- schemaErr0: `/message/catalog/bpp~1providers/0/items/5/category_id Invalid category ID found for item for on_search`
- bpp/descriptor/accept_bap_terms: `accept_bap_terms is not required in bpp/descriptor/tags for now`
- bpp/descriptor/collect_payment: `collect_payment is not required in bpp/descriptor/tags for now`
- prvdr0item0ff: `fulfillment_id in /bpp/providers[1]/items[1] should map to one of the fulfillments id in bpp/fulfillments`

## on_search_inc_refresh

- prvdr0item0ff: `fulfillment_id in /bpp/providers[1]/items[1] should map to one of the fulfillments id in bpp/fulfillments`

# Flow 2

## on_init

- schemaErr0: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'upi_address'`
- schemaErr1: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_ifsc_code'`
- schemaErr2: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_bank_account_no'`
- schemaErr3: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'bank_name'`
- schemaErr4: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'branch_name'`
- schemaErr5: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_type'`
- providertaxNumber: `provider_tax_number must be present for on_init`
- sttlmntcntrparty: `settlement_type is expected to be 'neft/rtgs/upi' in @ondc/org/settlement_details`

## confirm

- Same as above

## on_confirm

- Same as above
- schemaErr6: `/message/order/tags/1/list/0/value Value for tax_number must be a valid tax number i.e alphanumeric with 15 characters`
- message.order.fulfillments[c9211089-8b5e-4f41-9ee6-c69abd21d97b]: `'TAT' must be provided in message/order/fulfillments`
- message.order.tags[1].list: `np_type of on_search is not same to np_type of on_confirm`
- provider_tax_number: `provider_tax_number must be present for on_confirm`
- message/order/tags/bpp_terms: `List of bpp_terms mismatched in message/order/tags/bpp_terms for on_init and on_confirm`

## on_status_pending

- schemaErr0: `/message/order Invoice url must be present as part of documents objects (only in Order-picked-up state and thereafter)`
- schemaErr1: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_type'`
- on_status_pending_msgId: `Message id should not be same with previous calls`
- tmpstmp: `Created At timestamp for /on_status_pending should be equal to context timestamp at confirm`

## on_status_picked

- schemaErr0: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_type'`
- on_status_picked_msgId: `Message id should not be same with previous calls`
- tmpstmp: `Created At timestamp for /on_status_picked should be equal to context timestamp at confirm`

## on_status_out_for_delivery

- schemaErr0: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_type'`
- on_status_out_for_delivery_msgId: `Message id should not be same with previous calls`
- tmpstmp: `Created At timestamp for /on_status_out-for-delivery should be equal to context timestamp at confirm`

## on_status_delivered

- schemaErr0: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_type'`
- on_status_delivered_msgId: `Message id should not be same with previous calls`
- tmpstmp: `Created At timestamp for /on_status_delivered should be equal to context timestamp at confirm`

# Flow 4

## on_cancel

- orderCrtd: `order.created_at timestamp should match context.timestamp of confirm`

# Flow 5

## on_cancel

- schemaErr0: `/message/order/fulfillments/0/tags/1/list/0/code must be equal to one of the allowed values (reason_id,initiated_by,fulfillment_state,updated_at,retry_count,rto_id,id,currency,value,type)`
- schemaErr1: `/message/order/fulfillments/1/start/location must have required property 'id'`
- schemaErr2: `/message/order/fulfillments/1/start/location must have required property 'descriptor'`
- schemaErr3: `/message/order/fulfillments/1/end must have required property 'time'`
- schemaErr4: `/message/order/fulfillments/1/end must have required property 'person'`
- schemaErr5: `/message/order/fulfillments/1/end must have required property 'contact'`
- schemaErr6: `/message/order/fulfillments/1/end/location/address must have required property 'name'`
- schemaErr7: `/message/order/fulfillments/1/end/location/address must have required property 'building'`
- schemaErr8: `/message/order/fulfillments/1/end/location/address must have required property 'country'`
- on_cancel_msgId: `Message id should not be same with previous calls`

# Flow 6a

## on_update

- schemaErr0: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_ifsc_code'`
- schemaErr1: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_bank_account_no'`
- schemaErr2: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'bank_name'`
- schemaErr3: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'branch_name'`
- schemaErr4: `/message/order/payment/@ondc~1org~1settlement_details/0 must have required property 'settlement_type'`
- on_update_msgId: `Message id should not be same with previous calls`

## update

- Update settlement missing


# Flow 6b

## on_update_interim

- schemaErr0: `'payment' key should not have extra space in order.payment`
- schemaErr1: `/message/order/fulfillments/2/state/descriptor/code must be equal to one of the allowed values (Cancelled,Pending,Order-delivered,Return_Initiated,Return_Pick_Failed,Return_Approved,Return_Picked,Return_Delivered,Liquidated,Return_Rejected)

# flow 6c
## on_update_liquidated
- schemaErr0: /message/order must have required property 'payment'
- invldQuoteTrailPrices: quote_trail price and item quote price sum for on_update should be equal to the price as in on_confirm


# Please provide complete Flow 6 for next iteration , you can refer to the [contract](https://docs.google.com/document/d/1brvcltG_DagZ3kGr1ZZQk4hG4tze3zvcxmGV4NMTzr8/preview)

@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
@Jagannath-wits 

Flow 2: On-confirm
1. schemaErr6: /message/order/tags/1/list/0/value Value for tax_number must be a valid tax number i.e alphanumeric with 15 
characters ---> The tax number provided is a alphanumeric with 15 characters long, and /message/order/tags/1/list/0/value is related to Bap_terms which the tax number is sent by the buyer NP, we are the seller NP. please cross verify it. 
{
    "code": "tax_number",
    "value": "GSTIN1234567890"
}

2. message.order.tags[1].list: np_type of on_search is not same to np_type of on_confirm ---> Please cross verify this,                                
      - In the API contract you shared, message.order.tags[1].list does not contain np_type
      - and the np_type i.e 'ISN' it is same as in on_search and on_confirm message.order.tags[0].list

flow 5: on_cancel rto flow
. schemaErr1: /message/order/fulfillments/1/start/location must have required property 'id'
schemaErr2: /message/order/fulfillments/1/start/location must have required property 'descriptor'
schemaErr3: /message/order/fulfillments/1/end must have required property 'time'
schemaErr4: /message/order/fulfillments/1/end must have required property 'person'
schemaErr5: /message/order/fulfillments/1/end must have required property 'contact'
schemaErr6: /message/order/fulfillments/1/end/location/address must have required property 'name'
schemaErr7: /message/order/fulfillments/1/end/location/address must have required property 'building'
schemaErr8: /message/order/fulfillments/1/end/location/address must have required property 'country'

**Please cross verify this. It is same as in api contract. In api contract also for fulfillment[1] the above things are not there. and still if you feel it is not correct please provide more clarity on how to proceed with this test case. **
Please reply to the above queries ASAP

Thanks

**Comment by Jagannath-wits** [comment]
> @Jagannath-wits
> 
> Flow 2: On-confirm
> 
> 1. schemaErr6: /message/order/tags/1/list/0/value Value for tax_number must be a valid tax number i.e alphanumeric with 15
>    characters ---> The tax number provided is a alphanumeric with 15 characters long, and /message/order/tags/1/list/0/value is related to Bap_terms which the tax number is sent by the buyer NP, we are the seller NP. please cross verify it.
>    {
>    "code": "tax_number",
>    "value": "GSTIN1234567890"
>    }
> 2. message.order.tags[1].list: np_type of on_search is not same to np_type of on_confirm ---> Please cross verify this,
>    
>    * In the API contract you shared, message.order.tags[1].list does not contain np_type
>    * and the np_type i.e 'ISN' it is same as in on_search and on_confirm message.order.tags[0].list
> 
> flow 5: on_cancel rto flow . schemaErr1: /message/order/fulfillments/1/start/location must have required property 'id' schemaErr2: /message/order/fulfillments/1/start/location must have required property 'descriptor' schemaErr3: /message/order/fulfillments/1/end must have required property 'time' schemaErr4: /message/order/fulfillments/1/end must have required property 'person' schemaErr5: /message/order/fulfillments/1/end must have required property 'contact' schemaErr6: /message/order/fulfillments/1/end/location/address must have required property 'name' schemaErr7: /message/order/fulfillments/1/end/location/address must have required property 'building' schemaErr8: /message/order/fulfillments/1/end/location/address must have required property 'country'
> 
> **Please cross verify this. It is same as in api contract. In api contract also for fulfillment[1] the above things are not there. and still if you feel it is not correct please provide more clarity on how to proceed with this test case. ** Please reply to the above queries ASAP
> 
> Thanks

@SudarshanSirsi Please proceed with the other issues, and ignore these.

**Comment by SudarshanSirsi** [comment]
Hi @Jagannath-wits can you please review and give feedback on the above latest merge?

**Comment by NishthaMonga** [comment]
### Flow 1
## on_search_full_catalog_refresh
- Invalid item value: [512GB SSD Storage]. It must be one of the allowed values or match the regex pattern [^[0-9]{1,4}$].
- code should provided in /message/catalog/bpp/providers[0]/items[0]/descriptor

### Flow 2, Flow 3
## on_init
- Pan_id is different in tax_number and provider_tax_number in message.order.tags[0].list

## on_confirm
- List of bpp_terms mismatched in message/order/tags/bpp_terms for on_init and on_confirm here Code 'accept_bap_terms' present in first list but not in second list.

## on_status_pending, on_status_picked, on_status_out_for_delivery, on_status_delivered
- /start/range/start_time \"2024-05-03T06:08:28.465Z\" mismatched with /start/range/start_time \"2024-05-03T06:08:28.000Z\"

## on_status_packed logs missing

### Flow 4
## on_cancel
- Pre Cancel Updated at timeStamp is mandatory for on_cancel

### Flow 5
## on_cancel
- /message/order/fulfillments/1/start/location must have required property 'id'
- /message/order/fulfillments/1/start/location must have required property 'descriptor'
- /message/order/fulfillments/1/end must have required property 'time'
- /message/order/fulfillments/1/end/location/address must have required property 'name', 'building', 'country'
- Message id should not be same with previous calls
- precancel_state.updated_at of on_cancel is not equal with the on_status_out_for_delivery order.updated_at


@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
Hi @NishthaMonga 
**on_init**
-Pan_id is different in tax_number and provider_tax_number in message.order.tags[0].list
its because according to the ***[API Contract](https://docs.google.com/document/d/1brvcltG_DagZ3kGr1ZZQk4hG4tze3zvcxmGV4NMTzr8/edit)***  tax number refers to the GST number of the seller and provider_tax number refers to Pan number of the provider (page 283, footnote 717) . So according to the feedback if we should keep 'tax_number' & 'provider_tax_number' same, then which one we should use? whether GST tax number of the PAN number. because they both are not same for a company.

**on_confirm**
-List of bpp_terms mismatched in message/order/tags/bpp_terms for on_init and on_confirm here Code 'accept_bap_terms' present in first list but not in second list.

So for this should we remove accept_bap_terms from both the list of on_init and on_confirm or should I add 'accept_bap_terms' in on_init as well? Because according to api contract it is not there in on_init but it is present in on_confirm and it is mandatory according to api contract. ? what should we do ? 

**Flow 5**
on_cancel
-/message/order/fulfillments/1/start/location must have required property 'id'
-/message/order/fulfillments/1/start/location must have required property 'descriptor'
-/message/order/fulfillments/1/end must have required property 'time'
-/message/order/fulfillments/1/end/location/address must have required property 'name', 'building', 'country'

above issues are not correct, I have developed as in API contract, please cross verify this

Please reply to this ASAP
@Jagannath-wits @sandeepshahi you can also help here for 1st and 2nd one

Thank you

**Comment by Jagannath-wits** [comment]
> So according to the feedback if we should keep 'tax_number' & 'provider_tax_number' same, then which one we should use? whether GST tax number of the PAN number. because they both are not same for a company.

- Since you're an ISN as provided in on_search, the tax_number will be derived from provider_tax_number i.e after removing the 1st 2 and last 3 characters and if its same then it means that the seller NP and seller are same entity which further implies that it's an ISN; which is not the case in on_init.
{
                            "code": "tax_number",
                            "value": "29AAACZ6251B1ZE"
                        },
                        {
                            "code": "provider_tax_number",
                            "value": "AAACZ6251D"
                        }
                        
 > List of bpp_terms mismatched in message/order/tags/bpp_terms for on_init and on_confirm here Code 'accept_bap_terms' present in first list but not in second list.

- Accept_bap_terms is not required for now as static terms is not enabled yet, so just don't add Accept_bap_terms for now in any call and proceed.

> on_cancel
-/message/order/fulfillments/1/start/location must have required property 'id'
-/message/order/fulfillments/1/start/location must have required property 'descriptor'
-/message/order/fulfillments/1/end must have required property 'time'
-/message/order/fulfillments/1/end/location/address must have required property 'name', 'building', 'country'


- Please disregard the flow 5 errors and proceed.

@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
Got it now @Jagannath-wits , this clarified many things. Thank you for ur feedback

**Comment by sahil-ondc** [comment]
## Flow 1

### on_search (full catalog)

- /message/catalog/bpp/providers/items/descriptor/images should be publically accessible (`https://dgolwewdmdr5p.cloudfront.net/catalog/product/p/r/product_5490_1.png`)
- Variants should be provided
- order timing tags should be provided at provider level
- some items are having timestamps of 2021, why isn't those getting updated

## Flow 2

### on_status

- routing and tracking information should be provided

## Flow 3

### on_select (out-of-stock)

- error/message should be as per API contract
  `error.message in the response should have the list of corresponding item ids referenced in stringified object notation, e.g. if items "I1", "I2", "I3'' are out of stock, error.message would be encoded as: "[{\"item_id\":\"I1\",\"error\":\"40002\"},{\"item_id\":\"I2\",\"error\":\"40002\"},{\"item_id\":\"I3\",\"error\":\"40002\"}]". This is in addition to the error.code (40002) which is being sent by seller NP`

## Flow 5

### on_cancel (rto)

- in case of zero delivery it should not be provided in quote_trail

### on_status (rto-delivered, rto-disposed)

- rto-delivered or rto-disposed fulfillment state should be provided

## Flow 6

### on_status (after cancellation)

- @ondc/org/settlement_details should get updated after /update settlement call

### on_update (return)

- return type fulfillment should have pickup and delivery ranges

## Note

### on_search (full-catalog)

- /message/catalog/bpp/providers/items/descriptor short_desc and long_desc should not be same (catalog should be subset of production catalog)

@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
Hi @sahil-ondc 
***FLOW 1- On_search:***
-Variants should be provided
   We dont support varients. We only have simple products, thats the reason we are going without varients. 

***Flow 2 on_status:***
-routing and tracking information should be provided
     -may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; if tracking="true" & this is not 
       provided, default routing.type will be "P2P";
     - tracking attributes for fulfillment, may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true";
      This is what is there in API contract. since we dont have tracking as we have off-network logistic partner, i.e we are sending tracking=false so I dont thing we need to send this. what do you think?....and if we should send please tell what we need to send

***FLOW 5 - on_cancel(rto)***
     -in case of zero delivery it should not be provided in quote_trail
      Is this related to only to this flow or in every other flow, and all the othe API's as because in on_select, on_init, on_confirm, on_status, on_update in every api, delivery charge is 0, aprat from this which other locations we need to send like this?

***FLOW 6 - on_update***
    -return type fulfillment should have pickup and delivery ranges
    it is not there in api contract, should we go with this? 

Please update these in API contract as well. Because what I observed is many issues you mention are not specified in API contract, we will not be able to get to know until we submit the logs. So can you guys please include these in api contract as well, this will help the other participants as well. 

Please reply to this ASAP. 
Thank you

**Comment by sahil-ondc** [comment]
> Hi @sahil-ondc _**FLOW 1- On_search:**_ -Variants should be provided We dont support varients. We only have simple products, thats the reason we are going without varients.
> 
> _**Flow 2 on_status:**_ -routing and tracking information should be provided -may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; if tracking="true" & this is not provided, default routing.type will be "P2P"; - tracking attributes for fulfillment, may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; This is what is there in API contract. since we dont have tracking as we have off-network logistic partner, i.e we are sending tracking=false so I dont thing we need to send this. what do you think?....and if we should send please tell what we need to send
> 
> _**FLOW 5 - on_cancel(rto)**_ -in case of zero delivery it should not be provided in quote_trail Is this related to only to this flow or in every other flow, and all the othe API's as because in on_select, on_init, on_confirm, on_status, on_update in every api, delivery charge is 0, aprat from this which other locations we need to send like this?
> 
> _**FLOW 6 - on_update**_ -return type fulfillment should have pickup and delivery ranges it is not there in api contract, should we go with this?
> 
> Please update these in API contract as well. Because what I observed is many issues you mention are not specified in API contract, we will not be able to get to know until we submit the logs. So can you guys please include these in api contract as well, this will help the other participants as well.
> 
> Please reply to this ASAP. Thank you

@sandeepshahi Can you please provide clarity on these.

**Comment by sandeepshahi** [comment]
@SudarshanSirsi, please share the latest logs, let me check them.

Also, for the clarity of second part, there are certain features mentioned in the contract that are part of phase 2 i.e., static terms, offers, cancellation terms etc. and not yet enabled on the network. This information was already shared with all the network participants as part of [roll-out plan](https://docs.google.com/spreadsheets/d/1JZV6ZQzXcHUsOwegGtArX3DdIXYIy3gxkhQ00q7kICc/edit#gid=264161032) which also included test case scenarios. However, that phase has already passed, and NPs have begun implementing phase 2 features as well. Let's get this sorted.

**Comment by SudarshanSirsi** [comment]
Hi @sandeepshahi before submitting the logs can you also please clear the doubts I mentioned above?

> Hi @sahil-ondc _**FLOW 1- On_search:**_ -Variants should be provided We dont support varients. We only have simple products, thats the reason we are going without varients.
> 
> _**Flow 2 on_status:**_ -routing and tracking information should be provided -may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; if tracking="true" & this is not provided, default routing.type will be "P2P"; - tracking attributes for fulfillment, may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; This is what is there in API contract. since we dont have tracking as we have off-network logistic partner, i.e we are sending tracking=false so I dont thing we need to send this. what do you think?....and if we should send please tell what we need to send
> 
> _**FLOW 5 - on_cancel(rto)**_ -in case of zero delivery it should not be provided in quote_trail Is this related to only to this flow or in every other flow, and all the othe API's as because in on_select, on_init, on_confirm, on_status, on_update in every api, delivery charge is 0, aprat from this which other locations we need to send like this?
> 
> _**FLOW 6 - on_update**_ -return type fulfillment should have pickup and delivery ranges delivery range is not there in api contract, should we go with this?
> 
> Please update these in API contract as well. Because what I observed is many issues you mention are not specified in API contract, we will not be able to get to know until we submit the logs. So can you guys please include these in api contract as well, this will help the other participants as well.
> 
> Please reply to this ASAP. Thank you

**Comment by sandeepshahi** [comment]
> Hi @sahil-ondc _**FLOW 1- On_search:**_ -Variants should be provided We dont support varients. We only have simple products, thats the reason we are going without varients.

- All the features are mandatory; Independent SKUs are fine, but handling of variants must also be supported.

> _**Flow 2 on_status:**_ -routing and tracking information should be provided -may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; if tracking="true" & this is not provided, default routing.type will be "P2P"; - tracking attributes for fulfillment, may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; This is what is there in API contract. since we dont have tracking as we have off-network logistic partner, i.e we are sending tracking=false so I dont thing we need to send this. what do you think?....and if we should send please tell what we need to send

- if tracking is false, then tracking info need not be provided but routing must be present, since electronics products are mostly routed via Hub

> _**FLOW 5 - on_cancel(rto)**_ -in case of zero delivery it should not be provided in quote_trail Is this related to only to this flow or in every other flow, and all the othe API's as because in on_select, on_init, on_confirm, on_status, on_update in every api, delivery charge is 0, aprat from this which other locations we need to send like this?

- quote_trail applies only when there is a price adjustment in the quote/breakup and the updated information is provided in the newly created logical fulfillment (applicable for on_cancel, on_status, or on_update). Some of the APIs mentioned in the query are not applicable.

> _**FLOW 6 - on_update**_ -return type fulfillment should have pickup and delivery ranges it is not there in api contract, should we go with this?

- must be provided for reverseqc; as mentioned in the contract as well; pls go through the contract

> Please update these in API contract as well. Because what I observed is many issues you mention are not specified in API contract, we will not be able to get to know until we submit the logs. So can you guys please include these in api contract as well, this will help the other participants as well.

- can you specify which observation is not present in the API contract? Also, please review the issue boards of other NPs to check the common issues being reported. These issue boards are open source and publicly available for a reason.


@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
> > Hi @sahil-ondc _**FLOW 1- On_search:**_ -Variants should be provided We dont support varients. We only have simple products, thats the reason we are going without varients.
> 
> * All the features are mandatory; Independent SKUs are fine, but handling of variants must also be supported.

- We only have the independent sku's as of now, but we will support it in the future.

> 
> > _**Flow 2 on_status:**_ -routing and tracking information should be provided -may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; if tracking="true" & this is not provided, default routing.type will be "P2P"; - tracking attributes for fulfillment, may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; This is what is there in API contract. since we dont have tracking as we have off-network logistic partner, i.e we are sending tracking=false so I dont thing we need to send this. what do you think?....and if we should send please tell what we need to send
> 
> * if tracking is false, then tracking info need not be provided but routing must be present, since electronics products are mostly routed via Hub
> 
> > _**FLOW 5 - on_cancel(rto)**_ -in case of zero delivery it should not be provided in quote_trail Is this related to only to this flow or in every other flow, and all the othe API's as because in on_select, on_init, on_confirm, on_status, on_update in every api, delivery charge is 0, aprat from this which other locations we need to send like this?
> 
> * quote_trail applies only when there is a price adjustment in the quote/breakup and the updated information is provided in the newly created logical fulfillment (applicable for on_cancel, on_status, or on_update). Some of the APIs mentioned in the query are not applicable.
> 
> > _**FLOW 6 - on_update**_ -return type fulfillment should have pickup and delivery ranges it is not there in api contract, should we go with this?
> 
> * must be provided for reverseqc; as mentioned in the contract as well; pls go through the contract
> 
- In [API Cntract](https://docs.google.com/document/d/1brvcltG_DagZ3kGr1ZZQk4hG4tze3zvcxmGV4NMTzr8/edit#heading=h.8e58nknme8h) for return approved scenario only there is return pickup time range is there which is fulfillment.start.time.range , and I am not able to find the return deliver time ranges(fulfillment.end.time.range) in any of reverse qc flow. So shall we submit it like that as in contract or should we include delivery time range as well with next log

**Comment by SudarshanSirsi** [comment]
@sandeepshahi 
> 
> * can you specify which observation is not present in the API contract? Also, please review the issue boards of other NPs to check the common issues being reported. These issue boards are open source and publicly available for a reason.
> 
- rto-delivered or rto-disposed fulfillment state should be provided, :- 
  No where it is mentioned that we have to submit the logs for this, not in testcase scenario document, not in contract, and while verifying with log utility tool also, it did not asked these logs, for FYI I have seen the logs of some of the network participants who's logs are verified , they did not submitted logs for this, but their logs are verified already. That's why we did not submitted these logs,

**on_status**
- routing and tracking information should be provided
  I agree this was present in API contract, it says there, if tracking is false no need to implement, or exactly it says (tracking attributes for fulfillment, may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true";), why bring it up as issue here if it is specified in contract? we have specified in logs as tracking= false. 
also for routing information it is nowhere specified, it should be provided for electronics it just says (may be provided, on or before fulfillment state of "Agent-assigned", if tracking = "true"; if tracking="true" & this is not provided, default routing.type will be "P2P";) If you see it from the perspective of a new NP, who is integrating just now, it is so much confusing, we will not get to know what to do exactly

**on_update (return)**
- return type fulfillment should have pickup and delivery ranges: 
   I understand fulfillment should have pickup time ranges is there(fulfillment.start.time.range), but what about the delivery time ranges? it is not specified in contract, not given as error in utility tool as well. And we checked other NP's who are into electronics, they also did not submitted the logs with delivery time range, 

also I will tell you another issue
below are 2 images of api contract for on_search for electronics, you can see the item.descriptor.code in both screenshot and foot note, you can see in first it was optional(footnote 496) that was taken around mid April, so before that is was optional, and suddenly it become mandatory for electronics which you can see in second image(footnote 497) that was taken in first week of may. so in the roll out plan you provided previously, these small things are not updated, one can not read a 500 page document line by line every time we submit the logs, 
![Screenshot (216)_apr10](https://github.com/ONDC-Official/v1.2.0-logs/assets/121338271/8b6354e4-6b1e-48b5-b69d-ef53281fef86)

![Screenshot (215)may5](https://github.com/ONDC-Official/v1.2.0-logs/assets/121338271/d91770c0-5892-407b-99f9-305c474da014)

**Comment by sandeepshahi** [comment]
## Flow 1

**/on_search**

- why are placeholder values of gtin number provided?
- variants must be supported
- images are still not accessible

**/on_search (inc)**

- is the message_id unique for unsolicited push based incremental calls?
- what's the point of enabling the item if the inventory is not available? Also, disable should only be used for permanently disabling the item

## Flow 2

**/on_select**

- how can disabled item: "C587505WIN8" be processed for order? Why wasn't an error thrown or status update provided through incremental catalog?

**/on_confirm**

- pickup range should not overlap with context/timestamp; must always be future dated

**/on_status**

- is Agent-assigned state supported?

## Flow 3,4

- similar issues as above

## Flow 5

**/on_cancel (rto)**

- pickup timestamp should not be removed from cancelled forward shipment; audit trail must be maintained
- invalid attribute end/contact/email; wasn't even present in earlier calls
- industry standard for delivery attempts is of 3 retries; is initiating rto after just one attempt intended?

## Flow 6

**/on_update (part cancel)**

- invalid attribute end/contact/email; attribute owned by buyer apps should not be modified by SNP; applicable for all the APIs

**/on_status (picked)**

- invoice must be updated; cancelled SKUs are also included. Why is this happening?

**/on_update (return approved)**

- items should only be mapped to return fulfillment once the reverseqc is successful (ie return is picked); return might get rejected if the reverseqc fails
- quote and quote_trail should only be updated once the reverseqc is successful
- part cancellation related refund settlement details should not change
- return pickup range should not overlap with context/timestamp; must always be future dated


@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
@sandeepshahi 
/on_status - No agent assigned is not supported. Since we are going with off network logistics.

**Comment by SudarshanSirsi** [comment]
@sandeepshahi any updates on latest merge?

**Comment by sandeepshahi** [comment]
## Flow 1

Clarifications:

- Is ZRPL an ISN? Couldn't find any related entry on NP portal

**/on_search**

- GTIN values seem to be placeholder values
- pls share the screenshot of "variants" being displayed on any buyer app and their handling through your seller app
- avoid mentioning discounts or offers that cannot be applied, such as 'Use DISCOUNT2UPI to get an additional 2% off when paying by UPI,' since the prepaid payment is currently managed by BAP and this offer isn't available.

**/on_search (inc)**

- Since buyer NPs use timestamp at provider / location / item level to handle the race condition; the timestamps should be updated accordingly

## Flow 2

**/on_status**

- if the order is automatically accepted in the /on_confirm phase, there is no need to accept the order again through unsolicited /on_status
- Why is P2H2P being performed for hyperlocal delivery?

## Flow 5

**/status (rto delivered)**

- order/updated_at timestamp can't be future dated; must be <= context/timestamp

## Flow 6

**/on_update (return picked)**

- actual return pickup timestamp is not captured correctly; provided timestamp is even before the return is approved


@SudarshanSirsi

**Comment by SudarshanSirsi** [comment]
> ## Flow 1
> Clarifications:
> 
> * Is ZRPL an ISN? Couldn't find any related entry on NP portal
- Yes we are an ISN Seller app, we will update the info in the portal
> 
> **/on_search**
> 
> * GTIN values seem to be placeholder values
- GTIN Numbers are the actual GTIN numbers which are associated with the product. You can verify them in GS1 website
> * pls share the screenshot of "variants" being displayed on any buyer app and their handling through your seller app
For the screenshot part, We are doing testing with reference seller app, our payload seems correct (you might also have checked), which is according to the api contract. But it is not showing up. as variant. It might be a bug from reference buyerapp side. and regarding handling the variant's from seller side, it is handled as configurable product, as supported by any magento based system. 
> * avoid mentioning discounts or offers that cannot be applied, such as 'Use DISCOUNT2UPI to get an additional 2% off when paying by UPI,' since the prepaid payment is currently managed by BAP and this offer isn't available.
> 
> **/on_search (inc)**
> 
> * Since buyer NPs use timestamp at provider / location / item level to handle the race condition; the timestamps should be updated accordingly
in on_search in for item level changes are there. there is only one timestamp as by the api contract. that is bpp/providers/item/time/timestamp which should be updated only when the item is enabled or disabled(this is the feedback I got above). So and apart from this and context.tiemstamp I am not able to find any other timestamp, can you please elaborate this issue?
> 
> ## Flow 2
> **/on_status**
> 
> * if the order is automatically accepted in the /on_confirm phase, there is no need to accept the order again through unsolicited /on_status
1. So If I understood it correctly, do you mean to say, we don't need to send /on_status with orderstatus as 'Accepted' and fulfillment status as (Pending) ? 
2. But I am thinking it like, the logs should be submitted with on_status pending this is what instructed to us and the utility tool also asks for it, with on_status (pending). So do we not need to submit /on_status with accepted and pending? or do we need to send the confirm as 'created' and send the first status as 'accepted'?  
Too much confusion here. Need more clarification
> * Why is P2H2P being performed for hyperlocal delivery?
> 
> ## Flow 5
> **/status (rto delivered)**
> 
> * order/updated_at timestamp can't be future dated; must be <= context/timestamp
Can you please cross verify this once again I am seeing the updated_at timestamp is less then context.timestamp by 3ms. https://github.com/ONDC-Official/v1.2.0-logs/blob/master/Zeitgeist_Retail-Seller_App/Flow%205/on_status(rto_delivered).json
> 
> ## Flow 6
> **/on_update (return picked)**
> 
> * actual return pickup timestamp is not captured correctly; provided timestamp is even before the return is approved
> 
> @SudarshanSirsi

@sandeepshahi Need clarification regarding these ASAP. 
Thank you

**Comment by SudarshanSirsi** [comment]
Hi @sandeepshahi needed your inputs

**Comment by SudarshanSirsi** [comment]
Any updates of latest merge? @sandeepshahi

**Comment by BLR-0118** [comment]
Flow 1
====
1. full_catalog_on_search:
- special_feature needs to be in the list of enum [here](https://github.com/ONDC-Official/protocol-network-extension/blob/main/enums/retail/special_feature.json);
- colour_name should be added;

Flow 2
====
1. /on_select:
- quote doesn't have tax line item?

2. /on_confirm:
- curious how you can assign pickup & delivery slots when order isn't yet accepted?

**Comment by BLR-0118** [comment]
@SudarshanSirsi - pls have the above issues fixed & make an update here;
clearing logs for v1.2.0 (RET14), subject to above being fixed;

**Comment by SudarshanSirsi** [comment]
@BLR-0118 
> # Flow 1
> 1. full_catalog_on_search:
> 
> * special_feature needs to be in the list of enum [here](https://github.com/ONDC-Official/protocol-network-extension/blob/main/enums/retail/special_feature.json);
This is fixed.
> * colour_name should be added;
> 
I Have a doubt regarding this, Since in the api contract, it says regarding colour_name as this -> optional - required for colours not in standard "colour hex to name" mapping libraries; (P. No - 156, footnote 503). So we did not included this. 
But now I have added this as well.

> # Flow 2
> 1. /on_select:
> 
> * quote doesn't have tax line item?
> 
Is this mandatory? because according to API contract, we should not include tax line item, if tax value is '0' (P.No 199, footnote 570 -> tax on fulfillment level charges, to be included only if not 0). So we did not included this one.  So do we need to add this as well even if the tax value is 0?
> 2. /on_confirm:
> 
> * curious how you can assign pickup & delivery slots when order isn't yet accepted?
Have doubt here. need some clarification. This happened because of lack of understanding. In API contract This was not specified correctly, so I referred the logs of some other NP's who's logs are already verified and are live on network, all of those were using pickup and delivery slots in their /on_confirm call. 
Also for us, after getting the confirm, and once the order is created at our end sucessfully, then only we send this on_confirm response followed by /on_status with accepted. So we were able to add the slots in on_confirm only.

- So now according to my understanding after this feedback is that, we should not send the fulfillment.start.time(pickup slot) & fulfillment.end.time (Delivery slot) in /on_confirm payload, when the state is 'Created'. is this right? 
Need your go ahead for the above point.

**Comment by BLR-0118** [comment]
Flow 2
====
1. /on_select: in electronics, there should be a non-zero tax, better to make it explicit;
2. pickup & delivery slots can be assigned only after order is accepted by store;

**Comment by SudarshanSirsi** [comment]
Hi @BLR-0118 We have implemented above things and we have fixed those issues.

Thank you.
